server/src/main/java/org/elasticsearch/common/util/BigArrays.java

```diff
@@ -356,6 +356,88 @@ public T set(long index, T value) {
 
     }
 
+    abstract public static class BinarySearcher<T>{
+
+        /**
+         * @return a negative integer, zero, or a positive integer if the
+         * array's value at <code>index</code> is less than, equal to, or greater than <code>value</code>
+         */
+        abstract protected int compare(int index, T value);
```

Comment:
I wonder if you can avoid the `T` and the required allocation for it by making `compare` and `distance` "imply" the thing. Its kind of weird, but you could use it like:
```
private double searchFor;
private final BinarySearcher nearestCentroid = new BinarySearcher() {
  @Override
  protected in compare(int index) { Double.compareTo(searchFor, centroids.get(index)); }
  @Override
  protected in distance(int index) { Math.abs(searchFor - centroids.get(index)); }
};
```

It's pretty nasty compared to what you can do with C++'s templates, especially with the mutable state, but it lets us write all of this with only the one allocation for the centroid BinarySearcher. That way when you look for a new centroid you don't need to allocate at all.

Reply:
Also, if you've never seen this, it is a trip:
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/Integer.java#L986-L1044